runtime/Cpp/runtime/src/atn/ATNState.h

```diff
@@ -114,14 +114,20 @@ namespace atn {
 
     virtual bool isNonGreedyExitState() const;
     virtual std::string toString() const;
-    virtual ATNStateType getStateType() const = 0;
+
+    ATNStateType getStateType() const { return _stateType; }
+
+  protected:
+    explicit ATNState(ATNStateType stateType) : _stateType(stateType) {}
 
   private:
     /// Used to cache lookahead during parsing, not used during construction.
 
     misc::IntervalSet _nextTokenWithinRule;
     std::atomic<bool> _nextTokenUpdated { false };
 
+    ATNStateType _stateType;
```

Comment:
These are only set once in the constructor, so can be made `const`.
The advantage is that then the compiler will also make sure e.g. that there is no default constructor or assignment operators, so we make sure that there is no accidental expensive copies generated (so, for instance, ` ATNState() = delete;` would not be needed, as there can be no default constructor generated as there is at least one const member)

This is true for all the changes added in this PR, that all can be made const: `ATNStateType _stateType`, `LexerActionType actionType`, `PredictionContextType _contextType`, `SemanticContextType _contextType`, `TransitionType _transitionType` and `ParseTreeType _treeType`

Reply:
PredictionContextType cannot be, because ArrayPredictionContext needs to be moveable in another pull request. The rest are now const though.